PetaPoco/Core/PocoData.cs

```diff
@@ -140,16 +141,28 @@ public static PocoData ForObject(object obj, string primaryKeyName, IMapper defa
             var t = obj.GetType();
             if (t == typeof(System.Dynamic.ExpandoObject))
             {
-                var pd = new PocoData();
-                pd.TableInfo = new TableInfo();
-                pd.Columns = new Dictionary<string, PocoColumn>(StringComparer.OrdinalIgnoreCase);
-                pd.Columns.Add(primaryKeyName, new ExpandoColumn() { ColumnName = primaryKeyName });
-                pd.TableInfo.PrimaryKey = primaryKeyName;
-                pd.TableInfo.AutoIncrement = true;
+                var pd = new PocoData
+                {
+                    TableInfo = new TableInfo { AutoIncrement = true },
+                    Columns = new Dictionary<string, PocoColumn>(StringComparer.OrdinalIgnoreCase)
+                };
+
                 foreach (var col in (obj as IDictionary<string, object>).Keys)
                 {
-                    if (col != primaryKeyName)
-                        pd.Columns.Add(col, new ExpandoColumn() { ColumnName = col });
+                    pd.Columns.Add(col, new ExpandoColumn() { ColumnName = col });
+                }
+
+                //Ensure primary key is always in Columns list.
+                if (pd.Columns.TryGetValue(primaryKeyName, out PocoColumn column))
+                {
+                    //Prefer obj primary key name over primaryKeyName e.g. prefer obj.ID over "Id"
+                    pd.TableInfo.PrimaryKey = column.ColumnName;
```

Comment:
I still don't see the need for this. The user is responsible for passing in the correct value, in the correct case, for `primaryKeyName`.

Is the issue just that the "item with the same key" message is unhelpful? Then let's make a better error message.

Reply:
This change is only a more robust version of the original code.
Handling incorrect casing of the primary key name is not the responsibility of this method.

Reply:
I must be missing something. Isn't line 159 basically what you had in the other PR? Use the column name if it differs by case from `primaryKeyName`? That's what the PR description says.

Also, in what ways is this more robust?

Reply:
I'm going to try and elaborate on the "more robust" comment. The way things are in my head, don't always come out on paper correctly....lol.

As a whole, the method is responsible for providing a `PocoData` for the given type. For the happy path it can just return whatever comes back from the `ForType` method, but in the case of `ExpandoObejct`, the type is not useful.

The way I see it, there are two main objectives.
1. Create `Columns` from `((IDictionary<string, object>)obj).Keys`. Like a make shift `Type` if you want.
2. Ensure resulting `PocoData` instance contains the primary key in both `TableInfo.PrimaryKey` and `Columns`.

Objective 1
The closest we can get to a type is to look at the `Keys` within `obj`. which means `Keys` trump `primaryKeyName`. In this case, `primaryKeyName` is just a means to identify which `Key`, in `obj`, is the primary key, if it's there (cause there's no guarantee). So we should fill the `Columns` dictionary with `Keys` first. This avoids the complication of having to deal with duplicate key issues.

Objective 2
Since the `Columns` dictionary is not case sensitive, casing is not a factor in this method. It will be easy to figure out if we've already taken care of adding the primary key. `TryGetValue` will give us the `ExpandoColumn` if there's a key in `Columns` that matches the `primaryKeyName`. If we did not find a match, we have no choice but to use whatever the user passed as `primaryKeyName` to fill both `TableInfo.PrimaryKey` and `Columns`.

Hope this transfers my chain of thought for this change clearly and earn your approval.

Reply:
@asherber I'll have to circle back to it and find it, but in one of your comments you brought up some of the rationale for escaping incoming but not outgoing identifiers (column names specifically as I recall). There is actually a failing test currently (only for Oracle, as the other databases don't stress it that way), that may actually be due to (my) incorrect calling code. Will run it by you to be sure once I track down that comment and rerun the tests, but I'm fairly certain it relates to this.

Reply:
I don't think that there's a good reason to privilege the keys in the Expando, because we have no idea where that Expando came from. I think you're assuming it came from `db.Fetch<dynamic>()`, but that's not necessarily true.

Since the user has specified a value for `primaryKeyName`, we should assume he knows what he's talking about and use that. If he's wrong, there will be an exception at some point, and we shouldn't try to prevent that.

I do wonder if it's reasonable to change the guard clause (line 102 in the original) from `if (col != primaryKeyName)` to `if (!col.Equals(primaryKeyName, StringComparison.OrdinalIgnoreCase))`. That takes into account the case-insensitive dictionary and get rid of the "key already exists" error message, but then the library seems to throw a null reference exception further down the line.

Reply:
Oh, the NRE is just in the published library. I see that there's a fix for this in the current Oracle branch (null check on `pc.PropertyInfo`). So maybe just changing the guard clause is the better approach? Thoughts?

Reply:
Seems like it all boils down to trusting the user knows what he's doing/talking about. So the question becomes: "How is the returned PocoData instance useful to the user?"

Let's assume the user shares our sentiment that it's tedious to manually generate a dynamic object. So we can deduce that the whole object is either cased correctly or incorrectly, there is no middle ground. This means ALL keys must be treated the same way. There should be no discrimination between primary key or non-primary key. This is what my gut feeling is telling me.

I understand what you're saying about the uncertainty of where the dynamic object came from in the first place.
Let's say the object came from the user (manually generated) or some other code that auto generates it (not PP). We can safely assume that, if a key that resembles the primary key is found within the dynamic object, that the object has correctly cased keys. If no key resembles the primary key, we HAVE to assume the user knows what he's talking about and use exactly what he provided as-is.
We can also assume that the user will use this poco and the returned PocoData instance as a matching pair. 

PocoData is useful for a lot of things but, for the most part from what I could gather scanning through the code, the GetValue method lies at the core. For dynamic objects, the Columns dictionary is filled with ExpandoColumn instances, which overrides this method. If the ColumnName does not match a key in the target, it returns null.

```c#
public override object GetValue(object target)
{
    object val = null;
    (target as IDictionary<string, object>).TryGetValue(ColumnName, out val);
    return val;
}
```
Note that here the comparison is case sensitive.

Consider the following unit test:
```c#
[Theory]
[InlineData("ID", "4b8a4bb7-0104-4154-a294-6514f70f5a3e")]
[InlineData("id", "4b8a4bb7-0104-4154-a294-6514f70f5a3e")]
[InlineData("Id", "4b8a4bb7-0104-4154-a294-6514f70f5a3e")]
[InlineData("ID", null)]
[InlineData("id", null)]
[InlineData("Id", null)]
public void GetValue_GivenDynamicPoco_ShouldReturnCorrectValueWithNoSideEffects(string primaryKeyName, string primaryKeyValue)
{
    dynamic poco = new ExpandoObject();
    var dict = (IDictionary<string, object>)poco;
    var hasPrimaryKey = primaryKeyValue != null;

    //Simulate poco, consistently having ID (uppercase) primary key, irrespective of the user input, for the purposes of this test
    if (hasPrimaryKey) dict[primaryKeyName.ToUpper()] = Guid.Parse(primaryKeyValue);
    poco.CAPS = "CAPS";
    poco.TitleCase = "TitleCase";
    poco.lowercase = "lowecase";
    poco.snake_case = "snake_case";
    poco.Null = default(string);

    //Original code fails here -> ArgumentException: An item with the same key has already been added.
    PocoData pd = PocoData.ForObject(poco, primaryKeyName, new ConventionMapper());

    //This requires pd.TableInfo.PrimaryKey to be equal to pd.Columns[primaryKeyName].ColumnName
    //otherwise the primary key's key will be included in this list
    pd.UpdateColumns.ShouldBe(new[] { "CAPS", "TitleCase", "lowercase", "snake_case", "Null" });

    //This is not populated in case of dynamic objects
    pd.QueryColumns.ShouldBeNull();

    foreach (var column in pd.Columns)
    {
        var columnName = column.Value.ColumnName;

        //This will require ColumnName to be equal to poco Keys
        //if they differ GetValue will just silently return null
        object actual = column.Value.GetValue(poco);

        //We can't use dict[column.Value.ColumnName] so we'll test if it's there first
        if (dict.TryGetValue(columnName, out object expected))
            actual.ShouldBeSameAs(expected);

        if (hasPrimaryKey)
        {
            if (columnName == "Null") actual.ShouldBeNull();
            //Guard Clause fails here -> Shouldly.ShouldAssertException: actual should not be null but was
            else actual.ShouldNotBeNull();
        }
    }
}
```

Original Code
![image](https://github.com/CollaboratingPlatypus/PetaPoco/assets/298525/d33a4a08-7175-4856-921b-b3afde65e76c)
![image](https://github.com/CollaboratingPlatypus/PetaPoco/assets/298525/1e900ff4-71d7-430f-8935-3f78e34e28a5)
When the poco doesn't contain the primary key or if the user provides a primary key name with the correct case, the tests are happy. Otherwise we encounter the original issue, that sparked this PR, in PocoData on this line:
```c#
PocoData pd = PocoData.ForObject(poco, primaryKeyName, new ConventionMapper());
```

Guard Clause
![image](https://github.com/CollaboratingPlatypus/PetaPoco/assets/298525/8bd99794-6ba5-4a91-bfcb-894fabf9b87e)
![image](https://github.com/CollaboratingPlatypus/PetaPoco/assets/298525/d7120a28-25f0-47ef-9d1a-d730ffbb41f6)
When the poco doesn't contain the primary key or if the user provides a primary key name with the correct case, the tests are happy. Otherwise GetValue fails to return the expected value. The test fails on this line:
```c#
else actual.ShouldNotBeNull();
```

Proposed Implementation
![image](https://github.com/CollaboratingPlatypus/PetaPoco/assets/298525/46129363-a690-461f-8437-7fd0b4ac909b)
Whether poco contains primary key or not, all tests pass. PocoData exactly maps the dynamic object.

There might be other usages of the PocoData that this test does not cover that might or might not highlight issues with the discrimination between primary key and other columns approach. Although I strongly believe that the proposed implementation is a way more robust approach that also compliments the usage of this library as a whole.

Man! You're really making me work for this...hahahaha. But I'm not complaining, my passion for coding is too great. I actually enjoyed this "deep" thinking exercise.

Reply:
@Curlack I've read and considered your points, and I still don't agree with you. To me, it comes down to trusting what the user passes in, *especially* if it's a string. If the user says that the primary key of his table is named `id`, and he passes in (for whatever reason) a dynamic that doesn't have a property with that exact name, then he's going to have problems, and he deserves them.

I do think it's reasonable to make the guard clause in `ForObject()` case insensitive, to avoid the unhelpful "key already exists" message.

@Ste1io I think that [this line](https://github.com/CollaboratingPlatypus/PetaPoco/blob/53e928c6dc44819aa834683abcd108029f960b77/PetaPoco/Database.cs#L3057) in the `development` branch needs an additional null propagator after `PropertyInfo`, to match the check you added in `AddParameter()`.

If both of those changes are made, then given a table with columns `ID` and `code` in a case-sensitive database (I tested in Postgres), all 3 of these tests fail with the same message, which seems correct.

```csharp
// connection string details omitted
// Postgres error: column "id" does not exist

public class MyClass
{
    public int ID { get; set; }
    public string code { get; set; }
}

[Fact]
public void Concrete()
{
    using (var db = new Database())
    {
        var c = new MyClass() { ID = 7, code = "Bar" };
        db.Update("test", "id", c);
    }
}

[Fact]
public void Anonymous()
{
    using (var db = new Database())
    {
        var a  = new { ID = 7, code = "Foo" };
        db.Update("test", "id", a);
    }
}

[Fact]
public void Dynamic()
{    
    using (var db = new Database())
    {
        var d = db.Fetch<dynamic>("select * from test").First();
        db.Update("test", "id", d);
    }
}
```

Ultimately this is @pleb's call, but I would vote for this solution (make guard clause case-insensitive, add null propagation to guard clause in `Database.SetParameterProperties()` rather than the one in this PR.

Reply:
Off screen I believe you have a mapper that lowercases table and column names?
If that's the case you could pass "Test" or even "TEST" as the table name argument and it will still be able to do the update for you.
So a user with no knowledge about what's going on under the hood, would ask: "Why is it complaining when I get the primary key name case wrong, but it doesn't care about the casing I use in the table name?"

Granted I also thought that PocoData acted like a mediator between the poco and the code interacting with the database. At least from the dynamic objects perspective. Like an "auto correct" is you want.

But let's here what @pleb thinks.

Reply:
No mapper. The code above is as-is.

Reply:
Off course. Sorry. Meant to say "if you have a mapper that lowercases...".

Reply:
I'm beginning to see why you feel so string about the primaryKeyName trumping the obj.Keys.
In your tests, do you also get "Table doesn't exist" error if you case table name wrong?

Reply:
A mapper isn't going to have an effect, because of [this line of code](https://github.com/CollaboratingPlatypus/PetaPoco/blob/53e928c6dc44819aa834683abcd108029f960b77/PetaPoco/Database.cs#L2507). The `update` statement (in this case) will always use the literal primary key name passed in by the user.

I think the goal is to have dynamics work the same way as other classes; the changes I'm proposing for `dynamic` seem to accomplish that.

Yes, "table does not exist" if I use the wrong case for the table name. 

With the code in your PR, I do still get the "column does not exist" error, because I'm still passing in the wrong case in `primaryKeyName`, and that gets passed through untouched. But the PR changes other things in `PocoData.ForObject()` that I don't think need to be touched.

Reply:
Ok. One more thing. The side effect of `GetValue` silently returning null. Is that something we should be concerned about?

Reply:
> But the PR changes other things in `PocoData.ForObject()` that I don't think need to be touched.

I don't mean to be a pain, but could you please elaborate regarding the "things" you don't think need to be touched?

Reply:
Seems like it's intentional for `GetValue()` to return null if it can't find the column. I don't think there have been any complaints about this, so I'd be hesitant to change established behavior.

Regarding `ForObject()`, if the guard clause is the only thing that needs to be changed in that method, then it should be the only thing that gets changed. The code from line 153 on is no longer relevant, and while I usually like object initializers, I'm not sure that nested initializers are more readable than what's there currently. If you wanted to, you could move current lines 147 and 148 up under line 144, so the `TableInfo` stuff is together.

Reply:
This certainly appears to be one of those situations that calls for some explicit documentation for the xml doc comments, and probably a wiki page pointing out how PetaPoco handles case sensitivity when working with runtime objects. to avoid ambiguity. What exceptions are thrown, and when, might deserve some more consideration also, IMHO, as demonstrated in the two failing tests below.

While I share @asherber's concern that certain assumptions should not be made by the library, I think there is a use case where @Curlack's proposed changes would be beneficial. This may be a case of wanting to have my cake and eat it too, inside the current constraints, at least.

```cs
/*
PASS
*/
[Fact]
public virtual void Update_GivenTablePrimaryKeyNameAndDynamicType_ShouldBeValid()
{
    DB.Insert(_person);

    dynamic person = new System.Dynamic.ExpandoObject();
    person.Id = _person.Id;
    person.FullName = "Feta";
    person.Age = 19;
    person.Dob = new DateTime(1946, 1, 12, 5, 9, 4, DateTimeKind.Utc);
    person.Height = 190;

    ((int)DB.Update("People", "Id", (object)person)).ShouldBe(1);

    var personOther = DB.Single<Person>(_person.Id);
    personOther.ShouldNotBe(_person, true);
}

/*
Message:
  Shouldly.ShouldAssertException : `DB.Update("Note", "Id", (object)entity)`
    should not throw but threw
  System.ArgumentException
    with message
  "An item with the same key has already been added."

Stack Trace:
  UpdateTests.Update_GivenTablePrimaryKeyNameAndDynamicType_ShouldNotThrow() line 383
*/
[Fact]
[Trait("Issue", "#667")]
public virtual void Update_GivenTablePrimaryKeyNameAndDynamicType_ShouldNotThrow()
{
    var pd = PocoData.ForType(typeof(Note), DB.DefaultMapper);
    var tblNote = DB.Provider.EscapeTableName(pd.TableInfo.TableName);

    DB.Insert(_note);
    var entity = DB.Fetch<dynamic>($"SELECT * FROM {tblNote}").First();

    Should.NotThrow(() => DB.Update("Note", "Id", (object)entity));
}

/*
Message:
  Microsoft.CSharp.RuntimeBinder.RuntimeBinderException : 'System.Dynamic.ExpandoObject' does not contain a definition for 'Text'

Stack Trace:
  CallSite.Target(Closure, CallSite, Object )
  UpdateDelegates.UpdateAndExecute1[T0,TRet](CallSite site, T0 arg0)
  CallSite.Target(Closure, CallSite, Object )
  UpdateTests.Update_GivenTablePrimaryKeyNameAndDynamicType_ShouldUpdate() line 398
*/
[Fact]
public virtual void Update_GivenTablePrimaryKeyNameAndDynamicType_ShouldUpdate()
{
    var pd = PocoData.ForType(typeof(Note), DB.DefaultMapper);
    var tblNote = DB.Provider.EscapeTableName(pd.TableInfo.TableName);

    DB.Insert(_note);
    var entity = DB.Fetch<dynamic>($"SELECT * FROM {tblNote}").First();

    entity.Text += " was updated";

    DB.Update("Note", "Id", (object)entity);

    var entity2 = DB.Fetch<dynamic>($"SELECT * FROM {tblNote}").First();
    ((string)entity2.Text).ShouldContain("updated");
}
```

> @Ste1io I think that [this line](https://github.com/CollaboratingPlatypus/PetaPoco/blob/53e928c6dc44819aa834683abcd108029f960b77/PetaPoco/Database.cs#L3057) in the development branch needs an additional null propagator after PropertyInfo, to match the check you added in AddParameter().

@asherber I would agree. Want me to reopen/rebase the original bugfix pr and branch with that change or just patch in a fresh pr?

Reply:
I've reduce the `ForObject()` changes to just the guard clause, move the suggested lines of code and for readability added some whitespace.

Thanks for being so patient and responding so promptly to all my questions and comments.

I really don't want to delay this PR so I can finish the Oracle integration tests, so I'm conceding, but you can't say it was because of a lack of trying hahahaha.

Would be interesting to hear @pleb's view on this  though.

Reply:
> @Ste1io I think that [this line](https://github.com/CollaboratingPlatypus/PetaPoco/blob/53e928c6dc44819aa834683abcd108029f960b77/PetaPoco/Database.cs#L3057) in the development branch needs an additional null propagator after PropertyInfo, to match the check you added in AddParameter().

@asherber I would agree. Want me to reopen/rebase the original bugfix pr and branch with that change or just patch in a fresh pr?

Reply:
@Ste1io We can add the null propagator here; I was going to squash all of these anyway. 

I don't understand your comment above with test samples. Which commit are you running these against?

Reply:
> I don't understand your comment above with test samples. Which commit are you running these against?

The current public feature branch for oracle support #702, which iirc has no actual core changes currently in terms of caching or mapping behavior. The tests pass for all database providers except Oracle, which exposes the key-to-column caching behavior being discussed due to the transformations to unquoted identifiers applied by Oracle. I can point out in more detail if anything's not clear, as there are numerous things at play, but didn't want to inadvertently beat a dead horse.

Reply:
I think he's referring to [this](https://github.com/CollaboratingPlatypus/PetaPoco/blob/53e928c6dc44819aa834683abcd108029f960b77/PetaPoco/Database.cs#L2507) line. `primaryKeyName` is escaped, but the `value` provided [here](https://github.com/CollaboratingPlatypus/PetaPoco/blob/53e928c6dc44819aa834683abcd108029f960b77/PetaPoco/Database.cs#L2455) or [here](https://github.com/CollaboratingPlatypus/PetaPoco/blob/53e928c6dc44819aa834683abcd108029f960b77/PetaPoco/Database.cs#L2484) or [here](https://github.com/CollaboratingPlatypus/PetaPoco/blob/53e928c6dc44819aa834683abcd108029f960b77/PetaPoco/Database.cs#L2491) (depending on the scenario), is null (`GetValue()` side effect mentioned earlier).

The test is making a change to the poco and then updates the database. Then selects again from the database to confirm the update, but nothing changed cause the update affected 0 records. It affected 0 records because the statement correctly identifies "ID" as the primary key (escaping), but cannot determine the value cause "Id" was used to generate the `PocoData` instance.

Reply:
There's nothing wrong with the assertions, the exceptions occur during the identifier lookup while accessing the fetched POCO even though the POCO was created by the library:

> I don't think that there's a good reason to privilege the keys in the Expando, because we have no idea where that Expando came from. I think you're assuming it came from db.Fetch<dynamic>(), but that's not necessarily true.
>
> Since the user has specified a value for primaryKeyName, we should assume he knows what he's talking about and use that. If he's wrong, there will be an exception at some point, and we shouldn't try to prevent that.

Reply:
I would prefer not to hold this up for Oracle, since it shouldn't make anything worse for that db and may well make some things better. I suggest getting this merged and then seeing where we are.

Reply:
@Curlack Since this branch has 4 tiny commits, I'm going to ask you to do an interactive rebase to squash them down to one. If you run into trouble, or just don't want to get into this, I can take care of it for you. Which git tool are you currently using? CLI, or one of the GUIs?

Reply:
I'm using Visual Studio mostly, but switch to CLI when it doesn't support some command e.g. forcing a push like I did earlier.

Reply:
Should I include the null propogator also?

Reply:
@Curlack Let's continue the rebase conversation [here](https://gist.github.com/asherber/f8a452aeaf09bedca1806eb43631e9e8)